### PR TITLE
Update CHANGELOG.md for 3.16.0 release

### PR DESCRIPTION
**Pull request by fairclothjm**
Updates the changelog with items from the 3.16.0 milestone.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,19 @@
+## 3.16.0 (June 7, 2023)
+FEATURES:
+* VAULT-8099 LDAP secrets engine: ([#1859](https://github.com/hashicorp/terraform-provider-vault/pull/1859))
```
Comment (vinay-gopalan) [inline CHANGELOG.md:3]:
```suggestion
* Add support for LDAP secrets engine: ([#1859](https://github.com/hashicorp/terraform-provider-vault/pull/1859))
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+* Data source vault auth backends: ([#1827](https://github.com/hashicorp/terraform-provider-vault/pull/1827))
```
Comment (vinay-gopalan) [inline CHANGELOG.md:4]: nit
```suggestion
* Add new data source `vault_auth_backends`: ([#1827](https://github.com/hashicorp/terraform-provider-vault/pull/1827))
```

**Review by raymonstah — APPROVED** [review]
🎉

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,19 @@
...
+* Support allowed_domains_template on ssh_secret_backend_role. Fixes hashicorp#1675: ([#1676](https://github.com/hashicorp/terraform-provider-vault/pull/1676))
+
+IMPROVEMENTS:
+* Add support for retrying kv-v2 secret data writes: ([#1887](https://github.com/hashicorp/terraform-provider-vault/pull/1887))
+* Add back support for deriving the provider namespace from the Vault token's: ([#1841](https://github.com/hashicorp/terraform-provider-vault/pull/1841))
+
+BUGS:
+* check version for allowed_domains_template field: ([#1881](https://github.com/hashicorp/terraform-provider-vault/pull/1881))
```
Comment (vinay-gopalan) [inline CHANGELOG.md:12]: nit: Wondering if we can remove this, this was a fix for an unreleased change, so not sure if it warrants being on the changelog

**Review by vinay-gopalan — APPROVED** [review]
LGTM, couple of nits mostly

File: CHANGELOG.md (hunk 3 above)
Comment (fairclothjm) [reply CHANGELOG.md:12]: Yep, totally agree. Will update

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,19 @@
...
+* Fix DB engine password overwrite: ([#1876](https://github.com/hashicorp/terraform-provider-vault/pull/1876))
+* secrets/ldap: fix bindpass getting overwritten: ([#1875](https://github.com/hashicorp/terraform-provider-vault/pull/1875))
```
Comment (raymonstah) [inline CHANGELOG.md:14]: I wonder if @vinay-gopalan 's comment above applies for the LDAP bug fix here as well, since it was never officially released.

**Pull request merged**